File: app/src/main/java/com/worldventures/dreamtrips/modules/trips/command/GetTripsLocationsCommand.java
package com.worldventures.dreamtrips.modules.trips.command;

import com.worldventures.dreamtrips.R;
import com.worldventures.dreamtrips.api.trip.GetTripsLocationsHttpAction;
import com.worldventures.dreamtrips.api.trip.ImmutableGetTripsLocationsHttpAction;
import com.worldventures.dreamtrips.api.trip.model.TripPinWrapper;
import com.worldventures.dreamtrips.core.api.action.CommandWithError;
import com.worldventures.dreamtrips.core.janet.CommandActionBaseHelper;
import com.worldventures.dreamtrips.core.janet.JanetModule;
import com.worldventures.dreamtrips.core.janet.cache.CacheOptions;
import com.worldventures.dreamtrips.core.janet.cache.CachedAction;
import com.worldventures.dreamtrips.core.janet.cache.ImmutableCacheOptions;
import com.worldventures.dreamtrips.core.janet.dagger.InjectableAction;
import com.worldventures.dreamtrips.modules.trips.model.Pin;
import com.worldventures.dreamtrips.util.TripsFilterData;

import java.util.ArrayList;
import java.util.List;

import javax.inject.Inject;
import javax.inject.Named;

import io.techery.janet.ActionHolder;
import io.techery.janet.Janet;
import io.techery.janet.command.annotations.CommandAction;
import io.techery.mappery.MapperyContext;

@CommandAction
public class GetTripsLocationsCommand extends CommandWithError<List<Pin>> implements InjectableAction,
      CachedAction<List<Pin>> {

   @Inject Janet janet;
   @Inject MapperyContext mappery;

   private String searchQuery;
   private TripsFilterData tripsFilterData;

   private List<Pin> cachedResult;

   public GetTripsLocationsCommand(String searchQuery, TripsFilterData tripsFilterData) {
      this.searchQuery = searchQuery;
      this.tripsFilterData = tripsFilterData;
   }

   @Override
   protected void run(CommandCallback<List<Pin>> callback) throws Throwable {
      if (cachedResult != null && !cachedResult.isEmpty()) callback.onProgress(0);
      //
      janet.createPipe(GetTripsLocationsHttpAction.class)
            .createObservableResult(new GetTripsLocationsHttpAction(getParams()))
            .map(GetTripsLocationsHttpAction::response)
            .map(this::convert)
            .subscribe(CommandActionBaseHelper.ActionCommandSubscriber.wrap(callback));
   }

   public List<Pin> convert(List<TripPinWrapper> tripPins) {
      return mappery.convert(tripPins, Pin.class);
   }

   private GetTripsLocationsHttpAction.Params getParams() {
      if (tripsFilterData == null) return ImmutableGetTripsLocationsHttpAction.Params.builder().build();
      return ImmutableGetTripsLocationsHttpAction.Params.builder()
            .durationMax(tripsFilterData.getMaxNights())
            .durationMin(tripsFilterData.getMinNights())
            .priceMin(tripsFilterData.getMinPrice())
            .priceMax(tripsFilterData.getMaxPrice())
            .liked(tripsFilterData.isShowFavorites())
            .soldOut(tripsFilterData.isShowSoldOut())
            .recentFirst(tripsFilterData.isShowRecentlyAdded())
            .endDate(tripsFilterData.getEndDate())
            .startDate(tripsFilterData.getStartDate())
            .query(searchQuery)
            .activities(tripsFilterData.getAcceptedActivities())
            .regions(tripsFilterData.getAcceptedRegions())
            .build();
   }

   public List<Pin> getItems() {
      if (getResult() != null) return getResult();
      else return cachedResult;
   }

   @Override
   public int getFallbackErrorMessage() {
      return R.string.string_failed_to_load_trips;
   }

   @Override
   public void onRestore(ActionHolder holder, List<Pin> cache) {
      cachedResult = cache;
   }

   @Override
   public CacheOptions getCacheOptions() {
      return ImmutableCacheOptions.builder().build();
   }

   @Override
   public List<Pin> getCacheData() {
      return new ArrayList<>(getResult());
   }
}

File: app/src/main/java/com/worldventures/dreamtrips/modules/trips/service/command/GetTripsLocationsCommand.kt
package com.worldventures.dreamtrips.modules.trips.service.command

import com.worldventures.core.janet.CommandWithError
import com.worldventures.dreamtrips.R
import com.worldventures.dreamtrips.api.trip.GetTripsLocationsHttpAction
import com.worldventures.dreamtrips.api.trip.ImmutableGetTripsLocationsHttpAction
import com.worldventures.dreamtrips.modules.trips.model.filter.TripsFilterData
import com.worldventures.dreamtrips.modules.trips.model.map.Pin
import com.worldventures.janet.cache.CacheOptions
import com.worldventures.janet.cache.CachedAction
import com.worldventures.janet.injection.InjectableAction
import io.techery.janet.ActionHolder
import io.techery.janet.Command
import io.techery.janet.Janet
import io.techery.janet.command.annotations.CommandAction
import io.techery.mappery.MapperyContext
import java.util.ArrayList
import javax.inject.Inject

@CommandAction
class GetTripsLocationsCommand(private val searchQuery: String, private val tripsFilterData: TripsFilterData?) : CommandWithError<List<Pin>>(), InjectableAction, CachedAction<List<Pin>> {

   @Inject internal lateinit var janet: Janet
   @Inject internal lateinit var mappery: MapperyContext

   private var cachedResult: List<Pin> = emptyList()

   private val params: GetTripsLocationsHttpAction.Params
      get() = if (tripsFilterData == null) ImmutableGetTripsLocationsHttpAction.Params.builder().build()
      else ImmutableGetTripsLocationsHttpAction.Params.builder()
            .durationMax(tripsFilterData.maxNights)
            .durationMin(tripsFilterData.minNights)
            .priceMin(tripsFilterData.minPrice)
            .priceMax(tripsFilterData.maxPrice)
            .liked(tripsFilterData.isShowFavorites)
            .soldOut(tripsFilterData.isShowSoldOut)
            .recentFirst(tripsFilterData.isShowRecentlyAdded)
            .endDate(tripsFilterData.endDate)
            .startDate(tripsFilterData.startDate)
            .query(searchQuery)
            .activities(tripsFilterData.acceptedActivities)
            .regions(tripsFilterData.acceptedRegions)
            .build()

   val items: List<Pin>
      get() = result ?: cachedResult

   @Throws(Throwable::class)
   override fun run(callback: Command.CommandCallback<List<Pin>>) {
      if (!cachedResult.isEmpty()) callback.onProgress(0)
      janet.createPipe(GetTripsLocationsHttpAction::class.java)
            .createObservableResult(GetTripsLocationsHttpAction(params))
            .map { it.response() }
            .map { mappery.convert(it, Pin::class.java) }
            .subscribe(callback::onSuccess, callback::onFail)
   }

   override fun getFallbackErrorMessage() = R.string.string_failed_to_load_trips

   override fun onRestore(holder: ActionHolder<*>, cache: List<Pin>) {
      cachedResult = cache
   }

   override fun getCacheOptions() = CacheOptions()

   override fun getCacheData() = ArrayList(result)
}
